File: exercise/C#/day22/EID/StringExtensions.cs
using System.Text.RegularExpressions;
using LanguageExt;
using static LanguageExt.Option<int>;

namespace EID
{
    public static partial class StringExtensions
    {
        public static Option<int> ToInt(this string potentialNumber)
            => IsANumber(potentialNumber)
                ? int.Parse(potentialNumber)
                : None;

        private static bool IsANumber(string str) => NumberRegex().Match(str).Success;

        [GeneratedRegex("[0-9.]+")]
        private static partial Regex NumberRegex();
    }
}

File: exercise/java/day22/src/main/java/eid/StringExtensions.java
package eid;

import io.vavr.control.Option;
import lombok.experimental.UtilityClass;

import static io.vavr.control.Option.none;
import static io.vavr.control.Option.some;
import static java.lang.Integer.parseInt;

@UtilityClass
public class StringExtensions {
    public static Option<Integer> toInt(String potentialNumber) {
        return isANumber(potentialNumber) // Use Option<Integer> -> equivalent to Optional since java 8
                ? some(parseInt(potentialNumber))
                : none();
    }

    private static boolean isANumber(String str) {
        return str != null && str.matches("[0-9.]+");
    }
}